extensions/2.0/Khronos/KHR_meshopt_compression/README.md

```diff
@@ -0,0 +1,675 @@
+```
+
+Instead of using the raw index value, a zigzag-encoded 32-bit delta from `last` is used:
+
+```
+uint32_t decodeIndex(uint32_t v) {
+	int32_t delta = (v & 1) != 0 ? ~(v >> 1) : (v >> 1);
+
+	last += delta;
+	return last;
+}
+```
+
+The encoding for `code` is split into various cases, some of which are self-sufficient and some need to read extra data. The encoding is detailed below; after either path the triangle (a, b, c) is emitted to the output.
+
+- `0xX0`, where `X < 0xf`: Encodes a recently encountered edge and a `next` vertex.
+
+The edge (a, b) is read from the edge FIFO at index X (where 0 is the most recently added edge).
+The third index, `c`, is equal to `next` (which is then incremented).
+
+Edge (c, b) is pushed to the edge FIFO.
+Edge (a, c) is pushed to the edge FIFO.
+Vertex c is pushed to the vertex FIFO.
+
+- `0xXY`, where `X < 0xf` and `0 < Y < 0xd`: Encodes a recently encountered edge and a recently encountered vertex.
+
+The edge (a, b) is read from the edge FIFO at index X (where 0 is the most recently added edge).
+The third index, `c`, is read from the vertex FIFO at index Y (where 0 is the most recently added vertex; note that 0 is never actually read here, since `Y > 0`).
+
+Edge (c, b) is pushed to the edge FIFO.
+Edge (a, c) is pushed to the edge FIFO.
+
+- `0xXd` or `0xXe`, where `X < 0xf`: Encodes a recently encountered edge and a vertex that's adjacent to `last`.
+
+The edge (a, b) is read from the edge FIFO at index X (where 0 is the most recently added edge).
+The third index, `c`, is equal to `last-1` for `0xXd` and `last+1` for `0xXe`.
+
+`last` is set to `c` (effectively decrementing or incrementing it accordingly).
+
+Edge (c, b) is pushed to the edge FIFO.
+Edge (a, c) is pushed to the edge FIFO.
+Vertex c is pushed to the vertex FIFO.
+
+- `0xXf`, where `X < 0xf`: Encodes a recently encountered edge and a free-standing vertex encoded explicitly.
+
+The edge (a, b) is read from the edge FIFO at index X (where 0 is the most recently added edge).
+The third index, `c`, is decoded using `decodeIndex` by reading extra bytes from `data` (and also updates `last`).
+
+Edge (c, b) is pushed to edge FIFO.
+Edge (a, c) is pushed to edge FIFO.
+Vertex c is pushed to the vertex FIFO.
+
+- `0xfY`, where `Y < 0xe`: Encodes three indices using `codeaux` table lookup and vertex FIFO.
+
+The table `codeaux` is used to read the element Y; let's assume that results in `0xZW`.
+
+The first index, `a`, is equal to `next`; `next` is incremented to decode b/c correctly.
+The second index, `b`, is equal to `next` if `Z == 0` (`next` is then incremented), or is read from vertex FIFO at index `Z-1` (where 0 is the most recently added vertex).
+The third index, `c`, is equal to `next` if `W == 0` (`next` is then incremented), or is read from vertex FIFO at index `W-1` (where 0 is the most recently added vertex).
+
+Note that in the process `next` is incremented from 1 to 3 times depending on values of Z/W.
```

Comment:
Please organize these steps as a formal procedure so that the order and conditions are clear. For example:

> The triangle indices are set as follows:
> 1. Set the first index, `a`, to `next`.
> 2. Increment `next`.
> 3. If `Z` is equal to zero,
>    1. set the second index, `b`, to `next`;
>    2. increment `next`.
> 4. Else the second index, `b`, is read from the vertex FIFO at index `Z-1`.
> 5. If `W` is equal to zero,
>    1. set the third index, `c`, to `next`;
>    2. increment `next`.
> 6. Else the third index, `c`, is read from the vertex FIFO at index `W-1`.

While here:
- what happens if the FIFO entries at the indices Z or W have not been set?
- what happens if `next` reaches the `UINT_MAX`?

Reply:
Updated the steps. Re: `next`: I'm assuming you're asking about wraparound? Logically, `next` could reach `UINT_MAX` if one were to encode a sequence that has all indices from 0 to `UINT_MAX` organized as an index buffer, although the resulting index buffer would be invalid per glTF restart index requirements. After that, `next` would wrap around to 0 similarly to `next` restart. I think it would make sense to specify `next` increment as well as `last` adjustment as using 32-bit unsigned wrap around; for `next` it feels better to spell it out once before the description as it's incremented in multiple places in the algorithm, so I will adjust accordingly.

Reply:
FWIW, there's an extension that allows primitive restart index usage.